source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,173 @@
+---
+Title: Engine Dedicated CPUs
+Category: feature
+Authors: sabusale
+---
+
+# Engine Dedicated CPUs
+
+## Summary
+
+Once choosing dedicated CPU policy, The vCPUs will be strictly pinned to a set of host pCPUs
+(similarly to static CPU pinning). The set of pCPUs will be chosen to match the required guest CPU topology.
+
+Engine sets the pinning automatically when starting the VM as opposed to the manual pinning that is
+done by the user and is static.
+
+## Owner
+
+-   Name: Saif Abu Saleh (sabusale)
+-   Email: sabusale@redhat.com
+
+## Responsibilities
+
+VDSM:
+
+- Makes sure that dedicated VMs vCPUs stay exclusively pinned to the given pCPU
+
+Engine:
+- Holds cached available cpus based on host capabilities and pinned CPUs
+- Schedule VM to run on best satisfying host based on the above cached data structure
+
+
+
+## Backend
+
+1. Add the new policies to CPU Pinning Policy enum
+2. Save host capabilities CPU topology in vds_dynamic
+3. Hold hostToAvailableCpu data structure:
+    it will be on this form `Map<Integer,Map<Integer,Map<Integer,List<String>>>>`
+   
+    Host id -> Map of Socket id to Map of Core id to list of cpus
+    (assuming host id, core id, socket id is unique)
+
+    **Save/Update cached data structure:**    
+
+    * On updateVDSStatisticsData update the data structure with host capabilities
+    * On every start/stop/migrate/… VM update the shared CPU pinning data structure
+    with used pinning
+    * In case manual pinned cpus exist, they should be subtracted from hostToAvailableCpu 
```

Comment:
why? i think we should only subtract dedicated, manual is shared pool and VDSM should handle it AFAIK.

Reply:
@liranr23 yes but note that VDSM doesn't evacuate vCPUs that were manually pinned
example:
we manually pin vCPU-0 to pCPU-2
we cannot exclusively pin another vCPU exclusively to pCPU-2 because vCPU-0 won't move elsewhere

so if `available` refers here to "available for exclusive pinning" then pCPU-2 should be subtracted as well